### import libraries

In [227]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
% matplotlib inline
import warnings
warnings.filterwarnings('ignore')

### read data into pandas dataframe

In [228]:
df=pd.read_csv("/resources/data/train_recommend.csv")

In [229]:
df2=pd.read_csv("/resources/data/test_recommend.csv")

### understanding data

In [230]:
df.head()

CustomerID  InvoiceNo  Quantity    InvoiceDate  UnitPrice Country StockCode
0       27270      27270         7  01/12/10 8:26       2.55      PX   85123AY
1       27270      27270         7  01/12/10 8:26       3.39      PX    71053R
2       27270      27270         9  01/12/10 8:26       2.75      PX   84406BH
3       27270      27270         7  01/12/10 8:26       3.39      PX   84029GV
4       27270      27270         7  01/12/10 8:26       3.39      PX   84029EX

In [231]:
df.tail()

CustomerID  InvoiceNo  Quantity     InvoiceDate  UnitPrice Country  \
330570      296397     296397        14  09/12/11 12:21       1.45      PX   
330571      103599     103599         9  09/12/11 12:49       2.95      PX   
330572      103599     103599        28  09/12/11 12:49       1.25      PX   
330573      103599     103599        28  09/12/11 12:49       8.95      PX   
330574      103599     103599        12  09/12/11 12:49       7.08      PX   

       StockCode  
330570    23498F  
330571    22061J  
330572    23275I  
330573    21217W  
330574    20685H

In [232]:

df2.shape

(103097, 7)

### make copy of dataframes

In [233]:
df_original=df.copy()

In [234]:
df2_original=df2.copy()

In [235]:
df.columns

Index(['CustomerID', 'InvoiceNo', 'Quantity', 'InvoiceDate', 'UnitPrice',
       'Country', 'StockCode'],
      dtype='object')

In [236]:
print("Dataset shape is", df.shape)

Dataset shape is (330575, 7)


In [237]:
df2.columns

Index(['Country', 'CustomerID', 'InvoiceDate', 'InvoiceNo', 'Quantity',
       'StockCode', 'UnitPrice'],
      dtype='object')

In [238]:
print("Dataset shape is", df2.shape)

Dataset shape is (103097, 7)


In [239]:
df.dtypes

CustomerID       int64
InvoiceNo        int64
Quantity         int64
InvoiceDate     object
UnitPrice      float64
Country         object
StockCode       object
dtype: object

In [240]:
df2.dtypes

Country         object
CustomerID       int64
InvoiceDate     object
InvoiceNo        int64
Quantity         int64
StockCode       object
UnitPrice      float64
dtype: object

In [241]:
df.head()

CustomerID  InvoiceNo  Quantity    InvoiceDate  UnitPrice Country StockCode
0       27270      27270         7  01/12/10 8:26       2.55      PX   85123AY
1       27270      27270         7  01/12/10 8:26       3.39      PX    71053R
2       27270      27270         9  01/12/10 8:26       2.75      PX   84406BH
3       27270      27270         7  01/12/10 8:26       3.39      PX   84029GV
4       27270      27270         7  01/12/10 8:26       3.39      PX   84029EX

In [242]:
e=df.iloc[: ,0:2]

In [243]:
purchase_count=pd.crosstab(df['CustomerID'],df['StockCode'])
purchase_count.head()

StockCode   10002F  10080Q  10120M  10123CU  10124AW  10124GF  10125J  10133M  \
CustomerID                                                                      
0                0       0       0        0        0        0       0       0   
900              0       0       0        0        0        0       0       0   
1800             0       0       0        0        0        0       0       0   
2790             0       0       0        0        0        0       0       0   
4590             0       0       0        0        0        0       0       0   

StockCode   10135P  11001M   ...     90214OZ  90214PB  90214RZ  90214SF  \
CustomerID                   ...                                          
0                0       0   ...           0        0        0        0   
900              2       1   ...           0        0        0        0   
1800             0       0   ...           0        0        0        0   
2790             0       0   ...           0        0        0        0   
4590             0       0   ...           0        0        0        0   

StockCode   90214TT  90214UO  90214VD  90214WH  90214YL  90214ZJ  
CustomerID                                                        
0                 0        0        0        0        0        0  
900               0        0        0        0        0        0  
1800              0        0        0        0        0        0  
2790              0        0        0        0        0        0  
4590              0        0        0        0        0        0  

[5 rows x 3810 columns]

In [244]:
purchase_count2=pd.crosstab(df['StockCode'],df['CustomerID'])
purchase_count2.head()

CustomerID  0       900     1800    2790    4590    5400    6300    6390    \
StockCode                                                                    
10002F           0       0       0       0       0       0       0       0   
10080Q           0       0       0       0       0       0       0       0   
10120M           0       0       0       0       0       0       0       0   
10123CU          0       0       0       0       0       0       0       0   
10124AW          0       0       0       0       0       0       0       0   

CustomerID  7200    7290     ...    572625  573525  573615  574515  575325  \
StockCode                    ...                                             
10002F           0       0   ...         0       0       0       0       0   
10080Q           0       0   ...         0       1       0       0       0   
10120M           0       0   ...         0       0       0       0       0   
10123CU          0       0   ...         0       0       0       0       0   
10124AW          0       0   ...         0       0       0       0       0   

CustomerID  576315  577125  577215  578025  580815  
StockCode                                           
10002F           0       0       0       0       0  
10080Q           0       0       0       0       0  
10120M           0       0       0       0       0  
10123CU          0       0       0       0       0  
10124AW          0       0       0       0       0  

[5 rows x 972 columns]

In [245]:
p=purchase_count2.sum()

In [246]:
s=pd.DataFrame({'purchase_counting':p.values})
s

purchase_counting
0                  390
1                  826
2                  736
3                  393
4                  252
5                  399
6                  240
7                  638
8                  263
9                 1020
10                 526
11                1221
12                 810
13                1996
14                  31
15                 571
16                 259
17                 502
18                 400
19                 330
20                1264
21                 611
22                 725
23                 637
24                  52
25                 709
26                  16
27                 560
28                  41
29                   1
..                 ...
942                137
943                 75
944                152
945                 77
946                 52
947                196
948                112
949                250
950                  7
951                148
952                 53
953                291
954                 25
955                104
956                 57
957                 22
958                 78
959                170
960                 85
961                107
962                138
963                112
964                  7
965                 28
966                 14
967                 18
968                 25
969                 53
970                  3
971                 18

[972 rows x 1 columns]

In [247]:
df.columns

Index(['CustomerID', 'InvoiceNo', 'Quantity', 'InvoiceDate', 'UnitPrice',
       'Country', 'StockCode'],
      dtype='object')

# Now let's combine customer id, stockcode and purchase counting in one table

In [248]:
a=df.iloc[:,0:1]

In [249]:
b=df.iloc[:,6:]

In [250]:
c=a.join(b)

In [251]:
q=c.join(s)
q.head()

CustomerID StockCode  purchase_counting
0       27270   85123AY              390.0
1       27270    71053R              826.0
2       27270   84406BH              736.0
3       27270   84029GV              393.0
4       27270   84029EX              252.0

In [252]:
q.isnull().sum()

CustomerID                0
StockCode                 0
purchase_counting    329603
dtype: int64

In [253]:
r=q.dropna()

In [254]:
r.isnull().sum()

CustomerID           0
StockCode            0
purchase_counting    0
dtype: int64

In [255]:
r.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 972 entries, 0 to 971
Data columns (total 3 columns):
CustomerID           972 non-null int64
StockCode            972 non-null object
purchase_counting    972 non-null float64
dtypes: float64(1), int64(1), object(1)
memory usage: 30.4+ KB


In [256]:
r.tail()

CustomerID StockCode  purchase_counting
967      300897    20685H               18.0
968      300897    22842H               25.0
969      300897    22494U               53.0
970      300897    21260A                3.0
971      200898    20754G               18.0

In [257]:
r.shape

(972, 3)

### we will create purchased dummy which will show status of purchasing by 0 or 1 w.r.t. customer id and stockcode

In [258]:
def create_data_dummy(r):
    data_dummy = r.copy()
    data_dummy['purchase_dummy'] = 1
    return data_dummy
data_dummy = create_data_dummy(r)
data_dummy.head()

CustomerID StockCode  purchase_counting  purchase_dummy
0       27270   85123AY              390.0               1
1       27270    71053R              826.0               1
2       27270   84406BH              736.0               1
3       27270   84029GV              393.0               1
4       27270   84029EX              252.0               1

### we normalize purchase frequency of each item across users by first creating a user-item matrix as follows

In [259]:
data_matrix = pd.pivot_table(data_dummy, values='purchase_counting', index='CustomerID', columns='StockCode')
data_matrix.head()

StockCode   10002F  10125J  15044BI  15056BLH  15056ND  16237N  16238J  \
CustomerID                                                               
10890          NaN     NaN    303.0       NaN      NaN     NaN     NaN   
11790          NaN     NaN      NaN       NaN      NaN     NaN     NaN   
21780          NaN     NaN      NaN       NaN      NaN     NaN     NaN   
27270          NaN     NaN      NaN       NaN      NaN     NaN     NaN   
37260          NaN     NaN      NaN       NaN      NaN     NaN     NaN   

StockCode   16258AG  17021U  17091AE   ...     90059CW  90059EC  90129FL  \
CustomerID                             ...                                 
10890           NaN     NaN      NaN   ...         NaN      NaN      NaN   
11790           NaN     NaN      NaN   ...         NaN      NaN      NaN   
21780           NaN     NaN      NaN   ...         NaN      NaN      NaN   
27270           NaN     NaN      NaN   ...         NaN      NaN      NaN   
37260           NaN     NaN      NaN   ...         NaN      NaN      NaN   

StockCode   90185CC  90199CA  90200AL  90200BI  90200CM  90200DJ  90210BN  
CustomerID                                                                 
10890           NaN      NaN      NaN      NaN      NaN      NaN      NaN  
11790           NaN      NaN      NaN      NaN      NaN      NaN      NaN  
21780           NaN      NaN      NaN      NaN      NaN      NaN      NaN  
27270           NaN      NaN      NaN      NaN      NaN      NaN      NaN  
37260           NaN      NaN      NaN      NaN      NaN      NaN      NaN  

[5 rows x 584 columns]

In [260]:
data_matrix=data_matrix.fillna(0)
data_matrix.head()

StockCode   10002F  10125J  15044BI  15056BLH  15056ND  16237N  16238J  \
CustomerID                                                               
10890          0.0     0.0    303.0       0.0      0.0     0.0     0.0   
11790          0.0     0.0      0.0       0.0      0.0     0.0     0.0   
21780          0.0     0.0      0.0       0.0      0.0     0.0     0.0   
27270          0.0     0.0      0.0       0.0      0.0     0.0     0.0   
37260          0.0     0.0      0.0       0.0      0.0     0.0     0.0   

StockCode   16258AG  17021U  17091AE   ...     90059CW  90059EC  90129FL  \
CustomerID                             ...                                 
10890           0.0     0.0      0.0   ...         0.0      0.0      0.0   
11790           0.0     0.0      0.0   ...         0.0      0.0      0.0   
21780           0.0     0.0      0.0   ...         0.0      0.0      0.0   
27270           0.0     0.0      0.0   ...         0.0      0.0      0.0   
37260           0.0     0.0      0.0   ...         0.0      0.0      0.0   

StockCode   90185CC  90199CA  90200AL  90200BI  90200CM  90200DJ  90210BN  
CustomerID                                                                 
10890           0.0      0.0      0.0      0.0      0.0      0.0      0.0  
11790           0.0      0.0      0.0      0.0      0.0      0.0      0.0  
21780           0.0      0.0      0.0      0.0      0.0      0.0      0.0  
27270           0.0      0.0      0.0      0.0      0.0      0.0      0.0  
37260           0.0      0.0      0.0      0.0      0.0      0.0      0.0  

[5 rows x 584 columns]

In [261]:
data_matrix.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 52 entries, 10890 to 471825
Columns: 584 entries, 10002F to 90210BN
dtypes: float64(584)
memory usage: 237.7 KB


### it means there are 52 cutomers & 584 items in our dataset

### now we will normalize above dataset i.e. data_matrix

In [262]:
data_matrix_norm = (data_matrix-data_matrix.min())/(data_matrix.max()-data_matrix.min())
data_matrix_norm.head()

StockCode   10002F  10125J  15044BI  15056BLH  15056ND  16237N  16238J  \
CustomerID                                                               
10890          0.0     0.0      1.0       0.0      0.0     0.0     0.0   
11790          0.0     0.0      0.0       0.0      0.0     0.0     0.0   
21780          0.0     0.0      0.0       0.0      0.0     0.0     0.0   
27270          0.0     0.0      0.0       0.0      0.0     0.0     0.0   
37260          0.0     0.0      0.0       0.0      0.0     0.0     0.0   

StockCode   16258AG  17021U  17091AE   ...     90059CW  90059EC  90129FL  \
CustomerID                             ...                                 
10890           0.0     0.0      0.0   ...         0.0      0.0      0.0   
11790           0.0     0.0      0.0   ...         0.0      0.0      0.0   
21780           0.0     0.0      0.0   ...         0.0      0.0      0.0   
27270           0.0     0.0      0.0   ...         0.0      0.0      0.0   
37260           0.0     0.0      0.0   ...         0.0      0.0      0.0   

StockCode   90185CC  90199CA  90200AL  90200BI  90200CM  90200DJ  90210BN  
CustomerID                                                                 
10890           0.0      0.0      0.0      0.0      0.0      0.0      0.0  
11790           0.0      0.0      0.0      0.0      0.0      0.0      0.0  
21780           0.0      0.0      0.0      0.0      0.0      0.0      0.0  
27270           0.0      0.0      0.0      0.0      0.0      0.0      0.0  
37260           0.0      0.0      0.0      0.0      0.0      0.0      0.0  

[5 rows x 584 columns]

In [263]:
data_matrix_norm.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 52 entries, 10890 to 471825
Columns: 584 entries, 10002F to 90210BN
dtypes: float64(584)
memory usage: 237.7 KB


In [264]:
# create a table for input to the modeling  
z = data_matrix_norm.reset_index() 
z.index.names = ['scaled_purchase_freq'] 
data_norm = pd.melt(z, id_vars=['CustomerID'], value_name='scaled_purchase_freq').dropna()
print(data_norm.shape)
data_norm.head()

(30368, 3)


CustomerID StockCode  scaled_purchase_freq
0       10890    10002F                   0.0
1       11790    10002F                   0.0
2       21780    10002F                   0.0
3       27270    10002F                   0.0
4       37260    10002F                   0.0

In [265]:
data_norm.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 30368 entries, 0 to 30367
Data columns (total 3 columns):
CustomerID              30368 non-null int64
StockCode               30368 non-null object
scaled_purchase_freq    30368 non-null float64
dtypes: float64(1), int64(1), object(1)
memory usage: 949.0+ KB


In [266]:
data_norm.isnull().sum()

CustomerID              0
StockCode               0
scaled_purchase_freq    0
dtype: int64

In [267]:
data_norm['scaled_purchase_freq'].unique()

array([0.        , 1.        , 0.95446585, 0.48511384, 0.77816901,
       0.30273973, 0.38082192, 0.3109589 , 0.13636364, 0.70954357,
       0.33110976, 0.01879927, 0.52147239, 0.04792332, 0.23727185,
       0.56585366, 0.68181818, 0.36074271, 0.15119363, 0.07953064,
       0.28943937, 0.16613162, 0.30979133, 0.85970149, 0.00597015,
       0.0486618 , 0.61538462, 0.18032787, 0.83410853, 0.00343643,
       0.71477663, 0.68041237, 0.06185567, 0.0125    , 0.02222222,
       0.5       , 0.95737705, 0.47727273, 0.0619469 , 0.39085546,
       0.28293413, 0.20359281, 0.02844311, 0.04697987, 0.76510067,
       0.01362862, 0.27427598, 0.14736842, 0.35767511, 0.8852459 ,
       0.19374069, 0.28150573, 0.30769231, 0.18657938, 0.3207856 ,
       0.11598303, 0.01980198, 0.23681733, 0.45307443, 0.98412698,
       0.16931217, 0.16402116, 0.03557312, 0.70355731, 0.17431193,
       0.55152888, 0.06568516, 0.01485149, 0.08745875, 0.51980198,
       0.34818482, 0.04153355, 0.08945687, 0.05054945, 0.21796

### we have normalized their purchase history, from 0–1 (with 1 being the most number of purchase for an item and 0 being 0 purchase count for that item).

In [268]:
from sklearn.model_selection import train_test_split
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline

In [269]:
! pip install turicreate

In [270]:

import turicreate as tc
from turicreate import SFrame

In [271]:
def split_data(data):
    '''
    Splits dataset into training and test set.
    
    Args:
        data (pandas.DataFrame)
        
    Returns
        train_data (tc.SFrame)
        test_data (tc.SFrame)
    '''
    train, test = train_test_split(data, test_size = 0.2)
    train_data = tc.SFrame(train)
    test_data = tc.SFrame(test)
    return train_data, test_data

In [272]:
train_data, test_data = split_data(r)
train_data_dummy, test_data_dummy = split_data(data_dummy)
train_data_norm, test_data_norm = split_data(data_norm)

In [273]:
# constant variables to define field names include:
user_id = 'CustomerID'
item_id = 'StockCode'
users_to_recommend = list(r[user_id])
n_rec = 1 # number of items to recommend
n_display = 628 # to display the first few rows in an output dataset

In [274]:
def model(train_data, name, user_id, item_id, target, users_to_recommend, n_rec, n_display):
    if name == 'popularity':
        model = tc.popularity_recommender.create(train_data, 
                                                    user_id=user_id, 
                                                    item_id=item_id, 
                                                    target=target)
    elif name == 'cosine':
        model = tc.item_similarity_recommender.create(train_data, 
                                                    user_id=user_id, 
                                                    item_id=item_id, 
                                                    target=target, 
                                                    similarity_type='cosine')
    elif name == 'pearson':
        model = tc.item_similarity_recommender.create(train_data, 
                                                    user_id=user_id, 
                                                    item_id=item_id, 
                                                    target=target, 
                                                    similarity_type='pearson')
        
    recom = model.recommend(users=users_to_recommend, k=n_rec)
    recom.print_rows(n_display)
    return model

###  Popularity Model as Baseline
### The popularity model takes the most popular items for recommendation. 
### These items are products with the highest number of sells across customers.

### i. Using purchase count

In [275]:
name = 'popularity'
target = 'purchase_counting'
popularity = model(train_data, name, user_id, item_id, target, 
                   users_to_recommend, n_rec, n_display)

Preparing data set.

Data has 777 observations with 51 users and 501 items.

Data prepared in: 0.006566s

777 observations to process; with 501 unique items.

+------------+-----------+--------+------+
| CustomerID | StockCode | score  | rank |
+------------+-----------+--------+------+
|   27270    |   10002F  | 1996.0 |  1   |
|   27270    |   10002F  | 1996.0 |  1   |
|   27270    |   10002F  | 1996.0 |  1   |
|   27270    |   10002F  | 1996.0 |  1   |
|   27270    |   10002F  | 1996.0 |  1   |
|   27270    |   10002F  | 1996.0 |  1   |
|   27270    |   10002F  | 1996.0 |  1   |
|   427266   |   10002F  | 1996.0 |  1   |
|   462735   |   21824R  | 1610.0 |  1   |
|   462735   |   21824R  | 1610.0 |  1   |
|   462735   |   21824R  | 1610.0 |  1   |
|   462735   |   21824R  | 1610.0 |  1   |
|   462735   |   21824R  | 1610.0 |  1   |
|   462735   |   21824R  | 1610.0 |  1   |
|   462735   |   21824R  | 1610.0 |  1   |
|   462735   |   21824R  | 1610.0 |  1   |
|   462735   |   21824R  | 1610.0 |  1   |
|   462735   |   21824R  | 1610.0 |  1   |
|   462735   |   21824R  | 1610.0 |  1   |
|   462735   |   21824R  | 1610.0 |  1   |
|   462735 

# using purchase dummy

In [276]:
name = 'popularity'
target = 'purchase_dummy'
pop_dummy = model(train_data_dummy, name, user_id, item_id, target, 
                  users_to_recommend, n_rec, n_display)

Warning: Ignoring columns purchase_counting;

To use these columns in scoring predictions, use a model that allows the use of additional features.

Preparing data set.

Data has 777 observations with 52 users and 497 items.

Data prepared in: 0.040389s

777 observations to process; with 497 unique items.

+------------+-----------+-------+------+
| CustomerID | StockCode | score | rank |
+------------+-----------+-------+------+
|   27270    |   22593E  |  1.0  |  1   |
|   27270    |   22593E  |  1.0  |  1   |
|   27270    |   22593E  |  1.0  |  1   |
|   27270    |   22593E  |  1.0  |  1   |
|   27270    |   22593E  |  1.0  |  1   |
|   27270    |   22593E  |  1.0  |  1   |
|   27270    |   22593E  |  1.0  |  1   |
|   427266   |   22593E  |  1.0  |  1   |
|   462735   |   22593E  |  1.0  |  1   |
|   462735   |   22593E  |  1.0  |  1   |
|   462735   |   22593E  |  1.0  |  1   |
|   462735   |   22593E  |  1.0  |  1   |
|   462735   |   22593E  |  1.0  |  1   |
|   462735   |   22593E  |  1.0  |  1   |
|   462735   |   22593E  |  1.0  |  1   |
|   462735   |   22593E  |  1.0  |  1   |
|   462735   |   22593E  |  1.0  |  1   |
|   462735   |   22593E  |  1.0  |  1   |
|   462735   |   22593E  |  1.0  |  1   |
|   462735   |   22593E  |  1.0  |  1   |
|   462735   |   22593E  |  1.0  |

#  using scaled purchase frequency

In [277]:
name = 'popularity'
target = 'scaled_purchase_freq'
pop_norm = model(train_data_norm, name, user_id, item_id, target, users_to_recommend, n_rec, n_display)

Preparing data set.

Data has 24294 observations with 52 users and 584 items.

Data prepared in: 0.035043s

24294 observations to process; with 584 unique items.

+------------+-----------+----------------------+------+
| CustomerID | StockCode |        score         | rank |
+------------+-----------+----------------------+------+
|   27270    |   22865Y  | 0.061465657111952406 |  1   |
|   27270    |   22865Y  | 0.061465657111952406 |  1   |
|   27270    |   22865Y  | 0.061465657111952406 |  1   |
|   27270    |   22865Y  | 0.061465657111952406 |  1   |
|   27270    |   22865Y  | 0.061465657111952406 |  1   |
|   27270    |   22865Y  | 0.061465657111952406 |  1   |
|   27270    |   22865Y  | 0.061465657111952406 |  1   |
|   427266   |   22632O  | 0.08060033990170411  |  1   |
|   462735   |   22468F  | 0.059437967518026824 |  1   |
|   462735   |   22468F  | 0.059437967518026824 |  1   |
|   462735   |   22468F  | 0.059437967518026824 |  1   |
|   462735   |   22468F  | 0.059437967518026824 |  1   |
|   462735   |   22468F  | 0.059437967518026824 |  1   |
|   462735   |   22468F  | 0.059437967518026824 |  1   |
|   462735   |   22468F  | 0.05

In [278]:
r.groupby(by=item_id)['purchase_counting'].mean().sort_values(ascending=False).head(15)

StockCode
10002F     1996.0
21824R     1610.0
21586H     1470.0
21041U     1452.0
22059M     1440.0
22098N     1407.0
22652D     1285.0
22661W     1264.0
21823I     1248.0
21724T     1221.0
20726C     1179.0
21544W     1136.0
90200CM    1085.0
90059BZ    1084.0
21867K     1065.0
Name: purchase_counting, dtype: float64

### this is showing that products have been purchased on an average how many times by all customers.

### Collaborative Filtering Model

###  Cosine similarity to define similarity b/w items
### using vectors for similarity or dissimilarity

### i. Using purchase count

In [279]:
name = 'cosine'
target = 'purchase_counting'
cos = model(train_data, name, user_id, item_id, target, users_to_recommend, n_rec, n_display)

Preparing data set.

Data has 777 observations with 51 users and 501 items.

Data prepared in: 0.006552s

Training model from provided data.

Gathering per-item and per-user statistics.

+--------------------------------+------------+

| Elapsed Time (Item Statistics) | % Complete |

+--------------------------------+------------+

| 9.929ms                        | 100        |

+--------------------------------+------------+

Setting up lookup tables.

Processing data in one pass using dense lookup tables.

+-------------------------------------+------------------+-----------------+

| Elapsed Time (Constructing Lookups) | Total % Complete | Items Processed |

+-------------------------------------+------------------+-----------------+

| 21.16ms                             | 3.25             | 16              |

| 29.505ms                            | 100              | 501             |

+-------------------------------------+------------------+-----------------+

Finalizing lookup tables.

Generating candidate set for working with new users.

Finished training in 0.031703s

+------------+-----------+--------------------+------+
| CustomerID | StockCode |       score        | rank |
+------------+-----------+--------------------+------+
|   27270    |   22803S  | 241.51558470726013 |  1   |
|   27270    |   22803S  | 241.51558470726013 |  1   |
|   27270    |   22803S  | 241.51558470726013 |  1   |
|   27270    |   22803S  | 241.51558470726013 |  1   |
|   27270    |   22803S  | 241.51558470726013 |  1   |
|   27270    |   22803S  | 241.51558470726013 |  1   |
|   27270    |   22803S  | 241.51558470726013 |  1   |
|   427266   |   20750C  | 0.5177045452594757 |  1   |
|   462735   |   20750C  | 4135.343503784388  |  1   |
|   462735   |   20750C  | 4135.343503784388  |  1   |
|   462735   |   20750C  | 4135.343503784388  |  1   |
|   462735   |   20750C  | 4135.343503784388  |  1   |
|   462735   |   20750C  | 4135.343503784388  |  1   |
|   462735   |   20750C  | 4135.343503784388  |  1   |
|   462735   |   20750C  | 4135.343503784388  |  1   |
|   462735

### ii. Using purchase dummy

In [280]:
name = 'cosine'
target = 'purchase_dummy'
cos_dummy = model(train_data_dummy, name, user_id, item_id, target, users_to_recommend, n_rec, n_display)

Warning: Ignoring columns purchase_counting;

To use these columns in scoring predictions, use a model that allows the use of additional features.

Preparing data set.

Data has 777 observations with 52 users and 497 items.

Data prepared in: 0.007951s

Training model from provided data.

Gathering per-item and per-user statistics.

+--------------------------------+------------+

| Elapsed Time (Item Statistics) | % Complete |

+--------------------------------+------------+

| 1.172ms                        | 100        |

+--------------------------------+------------+

Setting up lookup tables.

Processing data in one pass using dense lookup tables.

+-------------------------------------+------------------+-----------------+

| Elapsed Time (Constructing Lookups) | Total % Complete | Items Processed |

+-------------------------------------+------------------+-----------------+

| 1.696ms                             | 10               | 49              |

| 30.88ms                             | 100              | 497             |

+-------------------------------------+------------------+-----------------+

Finalizing lookup tables.

Generating candidate set for working with new users.

Finished training in 0.033335s

+------------+-----------+---------------------+------+
| CustomerID | StockCode |        score        | rank |
+------------+-----------+---------------------+------+
|   27270    |   82482D  |  0.6989601016044616 |  1   |
|   27270    |   82482D  |  0.6989601016044616 |  1   |
|   27270    |   82482D  |  0.6989601016044616 |  1   |
|   27270    |   82482D  |  0.6989601016044616 |  1   |
|   27270    |   82482D  |  0.6989601016044616 |  1   |
|   27270    |   82482D  |  0.6989601016044616 |  1   |
|   27270    |   82482D  |  0.6989601016044616 |  1   |
|   427266   |   22593E  |         0.0         |  1   |
|   462735   |   21506J  | 0.13630345038005284 |  1   |
|   462735   |   21506J  | 0.13630345038005284 |  1   |
|   462735   |   21506J  | 0.13630345038005284 |  1   |
|   462735   |   21506J  | 0.13630345038005284 |  1   |
|   462735   |   21506J  | 0.13630345038005284 |  1   |
|   462735   |   21506J  | 0.13630345038005284 |  1   |
|   462735   |   21506J  | 0.13630345038005284 |

### iii. Using scaled purchase count

In [281]:
name = 'cosine' 
target = 'scaled_purchase_freq' 
cos_norm = model(train_data_norm, name, user_id, item_id, target, users_to_recommend, n_rec, n_display)

Preparing data set.

Data has 24294 observations with 52 users and 584 items.

Data prepared in: 0.03466s

Training model from provided data.

Gathering per-item and per-user statistics.

+--------------------------------+------------+

| Elapsed Time (Item Statistics) | % Complete |

+--------------------------------+------------+

| 2.775ms                        | 100        |

+--------------------------------+------------+

Setting up lookup tables.

Processing data in one pass using dense lookup tables.

+-------------------------------------+------------------+-----------------+

| Elapsed Time (Constructing Lookups) | Total % Complete | Items Processed |

+-------------------------------------+------------------+-----------------+

| 5.425ms                             | 1.75             | 10              |

| 75.114ms                            | 100              | 584             |

+-------------------------------------+------------------+-----------------+

Finalizing lookup tables.

Generating candidate set for working with new users.

Finished training in 0.116752s

+------------+-----------+------------------------+------+
| CustomerID | StockCode |         score          | rank |
+------------+-----------+------------------------+------+
|   27270    |  84029GV  |  0.003714522887101937  |  1   |
|   27270    |  84029GV  |  0.003714522887101937  |  1   |
|   27270    |  84029GV  |  0.003714522887101937  |  1   |
|   27270    |  84029GV  |  0.003714522887101937  |  1   |
|   27270    |  84029GV  |  0.003714522887101937  |  1   |
|   27270    |  84029GV  |  0.003714522887101937  |  1   |
|   27270    |  84029GV  |  0.003714522887101937  |  1   |
|   427266   |   21065W  |          0.0           |  1   |
|   462735   |   22271R  |  0.001780661856896419  |  1   |
|   462735   |   22271R  |  0.001780661856896419  |  1   |
|   462735   |   22271R  |  0.001780661856896419  |  1   |
|   462735   |   22271R  |  0.001780661856896419  |  1   |
|   462735   |   22271R  |  0.001780661856896419  |  1   |
|   462735   |   22271R  |  0.001780661856896419  |  1  

### Pearson similarity

### i. Using purchase count

In [282]:
name = 'pearson'
target = 'purchase_counting'
pear = model(train_data, name, user_id, item_id, target, users_to_recommend, n_rec, n_display)

Preparing data set.

Data has 777 observations with 51 users and 501 items.

Data prepared in: 0.007231s

Training model from provided data.

Gathering per-item and per-user statistics.

+--------------------------------+------------+

| Elapsed Time (Item Statistics) | % Complete |

+--------------------------------+------------+

| 17.56ms                        | 100        |

+--------------------------------+------------+

Setting up lookup tables.

Processing data in one pass using dense lookup tables.

+-------------------------------------+------------------+-----------------+

| Elapsed Time (Constructing Lookups) | Total % Complete | Items Processed |

+-------------------------------------+------------------+-----------------+

| 18.254ms                            | 3.25             | 16              |

| 30.46ms                             | 100              | 501             |

+-------------------------------------+------------------+-----------------+

Finalizing lookup tables.

Generating candidate set for working with new users.

Finished training in 0.032317s

+------------+-----------+--------+------+
| CustomerID | StockCode | score  | rank |
+------------+-----------+--------+------+
|   27270    |   10002F  | 1996.0 |  1   |
|   27270    |   10002F  | 1996.0 |  1   |
|   27270    |   10002F  | 1996.0 |  1   |
|   27270    |   10002F  | 1996.0 |  1   |
|   27270    |   10002F  | 1996.0 |  1   |
|   27270    |   10002F  | 1996.0 |  1   |
|   27270    |   10002F  | 1996.0 |  1   |
|   427266   |   10002F  | 1996.0 |  1   |
|   462735   |   21824R  | 1610.0 |  1   |
|   462735   |   21824R  | 1610.0 |  1   |
|   462735   |   21824R  | 1610.0 |  1   |
|   462735   |   21824R  | 1610.0 |  1   |
|   462735   |   21824R  | 1610.0 |  1   |
|   462735   |   21824R  | 1610.0 |  1   |
|   462735   |   21824R  | 1610.0 |  1   |
|   462735   |   21824R  | 1610.0 |  1   |
|   462735   |   21824R  | 1610.0 |  1   |
|   462735   |   21824R  | 1610.0 |  1   |
|   462735   |   21824R  | 1610.0 |  1   |
|   462735   |   21824R  | 1610.0 |  1   |
|   462735 

### ii. Using purchase dummy

In [283]:
name = 'pearson'
target = 'purchase_dummy'
pear_dummy = model(train_data_dummy, name, user_id, item_id, target, users_to_recommend, n_rec, n_display)

Warning: Ignoring columns purchase_counting;

To use these columns in scoring predictions, use a model that allows the use of additional features.

Preparing data set.

Data has 777 observations with 52 users and 497 items.

Data prepared in: 0.009549s

Training model from provided data.

Gathering per-item and per-user statistics.

+--------------------------------+------------+

| Elapsed Time (Item Statistics) | % Complete |

+--------------------------------+------------+

| 24.321ms                       | 100        |

+--------------------------------+------------+

Setting up lookup tables.

Processing data in one pass using dense lookup tables.

+-------------------------------------+------------------+-----------------+

| Elapsed Time (Constructing Lookups) | Total % Complete | Items Processed |

+-------------------------------------+------------------+-----------------+

| 25.445ms                            | 5.5              | 28              |

| 53.437ms                            | 100              | 497             |

+-------------------------------------+------------------+-----------------+

Finalizing lookup tables.

Generating candidate set for working with new users.

Finished training in 0.055107s

+------------+-----------+-------+------+
| CustomerID | StockCode | score | rank |
+------------+-----------+-------+------+
|   27270    |   22593E  |  0.0  |  1   |
|   27270    |   22593E  |  0.0  |  1   |
|   27270    |   22593E  |  0.0  |  1   |
|   27270    |   22593E  |  0.0  |  1   |
|   27270    |   22593E  |  0.0  |  1   |
|   27270    |   22593E  |  0.0  |  1   |
|   27270    |   22593E  |  0.0  |  1   |
|   427266   |   22593E  |  0.0  |  1   |
|   462735   |   22593E  |  0.0  |  1   |
|   462735   |   22593E  |  0.0  |  1   |
|   462735   |   22593E  |  0.0  |  1   |
|   462735   |   22593E  |  0.0  |  1   |
|   462735   |   22593E  |  0.0  |  1   |
|   462735   |   22593E  |  0.0  |  1   |
|   462735   |   22593E  |  0.0  |  1   |
|   462735   |   22593E  |  0.0  |  1   |
|   462735   |   22593E  |  0.0  |  1   |
|   462735   |   22593E  |  0.0  |  1   |
|   462735   |   22593E  |  0.0  |  1   |
|   462735   |   22593E  |  0.0  |  1   |
|   462735   |   22593E  |  0.0  |

### iii. Using scaled purchase count

In [284]:
name = 'pearson' 
target = 'scaled_purchase_freq' 
pear_norm = model(train_data_norm, name, user_id, item_id, target, users_to_recommend, n_rec, n_display)

Preparing data set.

Data has 24294 observations with 52 users and 584 items.

Data prepared in: 0.049843s

Training model from provided data.

Gathering per-item and per-user statistics.

+--------------------------------+------------+

| Elapsed Time (Item Statistics) | % Complete |

+--------------------------------+------------+

| 11.477ms                       | 100        |

+--------------------------------+------------+

Setting up lookup tables.

Processing data in one pass using dense lookup tables.

+-------------------------------------+------------------+-----------------+

| Elapsed Time (Constructing Lookups) | Total % Complete | Items Processed |

+-------------------------------------+------------------+-----------------+

| 19.453ms                            | 1.75             | 10              |

| 1.87s                               | 100              | 584             |

+-------------------------------------+------------------+-----------------+

Finalizing lookup tables.

Generating candidate set for working with new users.

Finished training in 1.88339s

+------------+-----------+----------------------+------+
| CustomerID | StockCode |        score         | rank |
+------------+-----------+----------------------+------+
|   27270    |   22865Y  | 0.06024377866592663  |  1   |
|   27270    |   22865Y  | 0.06024377866592663  |  1   |
|   27270    |   22865Y  | 0.06024377866592663  |  1   |
|   27270    |   22865Y  | 0.06024377866592663  |  1   |
|   27270    |   22865Y  | 0.06024377866592663  |  1   |
|   27270    |   22865Y  | 0.06024377866592663  |  1   |
|   27270    |   22865Y  | 0.06024377866592663  |  1   |
|   427266   |   22632O  | 0.07979667272162247  |  1   |
|   462735   |   22468F  | 0.05876913298520051  |  1   |
|   462735   |   22468F  | 0.05876913298520051  |  1   |
|   462735   |   22468F  | 0.05876913298520051  |  1   |
|   462735   |   22468F  | 0.05876913298520051  |  1   |
|   462735   |   22468F  | 0.05876913298520051  |  1   |
|   462735   |   22468F  | 0.05876913298520051  |  1   |
|   462735   |   22468F  | 0.05

### Model Evaluation

### For evaluating recommendation engines, we can use the concept of RMSE and precision-recall.

### i. RMSE (Root Mean Squared Errors)
### ii. Recall
### iii. Precision

In [285]:
models_w_counts = [popularity, cos, pear]
models_w_dummy = [pop_dummy, cos_dummy, pear_dummy]
models_w_norm = [pop_norm, cos_norm, pear_norm]
names_w_counts = ['Popularity Model on Purchase Counts', 'Cosine Similarity on Purchase Counts', 'Pearson Similarity on Purchase Counts']
names_w_dummy = ['Popularity Model on Purchase Dummy', 'Cosine Similarity on Purchase Dummy', 'Pearson Similarity on Purchase Dummy']
names_w_norm = ['Popularity Model on Scaled Purchase Counts', 'Cosine Similarity on Scaled Purchase Counts', 'Pearson Similarity on Scaled Purchase Counts']

### Lets compare all the models we have built based on RMSE and precision-recall characteristics:

In [286]:
eval_counts = tc.recommender.util.compare_models(test_data, models_w_counts, model_names=names_w_counts)
eval_dummy = tc.recommender.util.compare_models(test_data_dummy, models_w_dummy, model_names=names_w_dummy)
eval_norm = tc.recommender.util.compare_models(test_data_norm, models_w_norm, model_names=names_w_norm)

PROGRESS: Evaluate model Popularity Model on Purchase Counts

Precision and recall summary statistics by cutoff
+--------+----------------------+-----------------------+
| cutoff |     mean_recall      |     mean_precision    |
+--------+----------------------+-----------------------+
|   1    |         0.0          |          0.0          |
|   2    |         0.0          |          0.0          |
|   3    | 0.005882352941176472 |  0.009803921568627453 |
|   4    | 0.005882352941176472 |  0.007352941176470588 |
|   5    | 0.005882352941176472 |  0.005882352941176472 |
|   6    | 0.005882352941176472 |  0.004901960784313726 |
|   7    | 0.005882352941176472 |  0.004201680672268907 |
|   8    | 0.005882352941176472 |  0.003676470588235294 |
|   9    | 0.005882352941176472 | 0.0032679738562091504 |
|   10   | 0.005882352941176472 |  0.002941176470588236 |
+--------+----------------------+-----------------------+
[10 rows x 3 columns]


Overall RMSE: 366.7267741315597

Per User RMSE (best

### RMSE 
### 1. POPULARITY: a) 359.88, b)0.0, c) 0.14
### 1. COSINE: a) 442.38, b)0.92, c) 0.14
### 1. PEARSON: a) 426.46, b)1.0, c) 0.14

###  Evaluation Summary
### We can see that the collaborative filtering algorithms work better than popularity model for purchase counts. Indeed, popularity model doesn’t give any personalizations as it only gives the same list of recommended items to every user.
### Precision and recall: Looking at the summary above, we see that the precision and recall for Purchase Counts > Purchase Dummy > Normalized Purchase Counts. However, because the recommendation scores for the normalized purchase data is zero and constant, we choose the dummy. In fact, the RMSE isn’t much different between models on the dummy and those on the normalized data.
### Since RMSE is higher using pearson distance thancosine, we would choose model the smaller mean squared errors, which in this case would be cosine.

### Therefore, we select the Cosine similarity on Purchase Dummy approach as our final model.


### final Output
### Finally, we would like to manipulate format for recommendation output to one we can export to csv, and also a function that will return recommendation list given a customer ID.

### We need to first rerun the model using the whole dataset, as we came to a final model using train data and evaluated with test set.



In [287]:
features=['CustomerID', 'InvoiceNo', 'Quantity', 'InvoiceDate', 'UnitPrice'] 

In [288]:
x=df[features]

In [289]:
y=df['StockCode']

In [290]:
#sklearn.metrics.fbeta_score(y, y_pred, beta, labels=None, pos_label=1, average=binary, sample_weight=None)

In [291]:
final_model = tc.item_similarity_recommender.create(tc.SFrame(data_dummy),  user_id=user_id, item_id=item_id, target='purchase_dummy', similarity_type='cosine')
recom = final_model.recommend(users=users_to_recommend, k=n_rec)
recom.print_rows(n_display)

Warning: Ignoring columns purchase_counting;

To use these columns in scoring predictions, use a model that allows the use of additional features.

Preparing data set.

Data has 972 observations with 52 users and 584 items.

Data prepared in: 0.008726s

Training model from provided data.

Gathering per-item and per-user statistics.

+--------------------------------+------------+

| Elapsed Time (Item Statistics) | % Complete |

+--------------------------------+------------+

| 1.407ms                        | 100        |

+--------------------------------+------------+

Setting up lookup tables.

Processing data in one pass using dense lookup tables.

+-------------------------------------+------------------+-----------------+

| Elapsed Time (Constructing Lookups) | Total % Complete | Items Processed |

+-------------------------------------+------------------+-----------------+

| 2.433ms                             | 0                | 0               |

| 56.931ms                            | 100              | 584             |

+-------------------------------------+------------------+-----------------+

Finalizing lookup tables.

Generating candidate set for working with new users.

Finished training in 0.058763s

+------------+-----------+----------------------+------+
| CustomerID | StockCode |        score         | rank |
+------------+-----------+----------------------+------+
|   27270    |   21871Y  |  0.8056334682873317  |  1   |
|   27270    |   21871Y  |  0.8056334682873317  |  1   |
|   27270    |   21871Y  |  0.8056334682873317  |  1   |
|   27270    |   21871Y  |  0.8056334682873317  |  1   |
|   27270    |   21871Y  |  0.8056334682873317  |  1   |
|   27270    |   21871Y  |  0.8056334682873317  |  1   |
|   27270    |   21871Y  |  0.8056334682873317  |  1   |
|   427266   |  85123AY  |         0.0          |  1   |
|   462735   |   22077P  |  0.1358635613792821  |  1   |
|   462735   |   22077P  |  0.1358635613792821  |  1   |
|   462735   |   22077P  |  0.1358635613792821  |  1   |
|   462735   |   22077P  |  0.1358635613792821  |  1   |
|   462735   |   22077P  |  0.1358635613792821  |  1   |
|   462735   |   22077P  |  0.1358635613792821  |  1   |
|   462735   |   22077P  |  0.1

### CSV output file
### Here we want to manipulate our result to a csv output. Let’s see what we have:

In [292]:
df_rec = recom.to_dataframe()
print(df_rec.shape)
df_rec.tail(50)

(972, 4)


CustomerID StockCode     score  rank
922      300897   35591TW  0.086427     1
923      300897   35591TW  0.086427     1
924      300897   35591TW  0.086427     1
925      300897   35591TW  0.086427     1
926      300897   35591TW  0.086427     1
927      300897   35591TW  0.086427     1
928      300897   35591TW  0.086427     1
929      300897   35591TW  0.086427     1
930      300897   35591TW  0.086427     1
931      300897   35591TW  0.086427     1
932      300897   35591TW  0.086427     1
933      300897   35591TW  0.086427     1
934      300897   35591TW  0.086427     1
935      300897   35591TW  0.086427     1
936      300897   35591TW  0.086427     1
937      300897   35591TW  0.086427     1
938      300897   35591TW  0.086427     1
939      300897   35591TW  0.086427     1
940      300897   35591TW  0.086427     1
941      300897   35591TW  0.086427     1
942      300897   35591TW  0.086427     1
943      300897   35591TW  0.086427     1
944      300897   35591TW  0.086427     1
945      300897   35591TW  0.086427     1
946      300897   35591TW  0.086427     1
947      300897   35591TW  0.086427     1
948      300897   35591TW  0.086427     1
949      300897   35591TW  0.086427     1
950      300897   35591TW  0.086427     1
951      300897   35591TW  0.086427     1
952      300897   35591TW  0.086427     1
953      300897   35591TW  0.086427     1
954      300897   35591TW  0.086427     1
955      300897   35591TW  0.086427     1
956      300897   35591TW  0.086427     1
957      300897   35591TW  0.086427     1
958      300897   35591TW  0.086427     1
959      300897   35591TW  0.086427     1
960      300897   35591TW  0.086427     1
961      300897   35591TW  0.086427     1
962      300897   35591TW  0.086427     1
963      300897   35591TW  0.086427     1
964      300897   35591TW  0.086427     1
965      300897   35591TW  0.086427     1
966      300897   35591TW  0.086427     1
967      300897   35591TW  0.086427     1
968      300897   35591TW  0.086427     1
969      300897   35591TW  0.086427     1
970      300897   35591TW  0.086427     1
971      200898   85123AY  0.000000     1

In [301]:
def create_output(model, users_to_recommend, n_rec, print_csv=True):
    recomendation = model.recommend(users=users_to_recommend, k=n_rec)
    df_rec = recomendation.to_dataframe()
    df_rec['recommendedProducts'] = df_rec.groupby([user_id])[item_id] \
        .transform(lambda x: '|'.join(x.astype(str)))
    df_output = df_rec[['CustomerID', 'recommendedProducts']]\
        .sort_values('CustomerID').set_index('CustomerID')
    if print_csv:
        df_output.to_csv('sub.csv')
        print("An output file can be found in 'output' folder with name 'option1_recommendation.csv'")
    return df_output

In [302]:
df_output = create_output(pear_norm, users_to_recommend, n_rec, print_csv=True)
print(df_output.shape)


An output file can be found in 'output' folder with name 'option1_recommendation.csv'
(972, 1)


In [303]:
df_output.head()

recommendedProducts
CustomerID                                                   
10890       84029EX|84029EX|84029EX|84029EX|84029EX|84029E...
10890       84029EX|84029EX|84029EX|84029EX|84029EX|84029E...
10890       84029EX|84029EX|84029EX|84029EX|84029EX|84029E...
10890       84029EX|84029EX|84029EX|84029EX|84029EX|84029E...
10890       84029EX|84029EX|84029EX|84029EX|84029EX|84029E...

In [304]:
df_output.shape

(972, 1)

In [305]:
from IPython.display import HTML
import base64

def create_download_link( submission, title = "Download CSV file", filename = "submission.csv"):  
    csv = submission.to_csv()
    b64 = base64.b64encode(csv.encode())
    payload = b64.decode()
    html = '<a download="{filename}" href="data:text/csv;base64,{payload}" target="_blank">{title}</a>'
    html = html.format(payload=payload,title=title,filename=filename)
    return HTML(html)

create_download_link(submission)